### DISCUSSION TITLE
弹出层问题，怎么自动调整高度？

### DISCUSSION THREAD

**Discussion by hulang** [layui/layui#2159] [问答]
### 议题条件

- [X] 我确认已查看官方使用文档：https://layui.dev ，但没有找到相关解决方案。
- [X] 我确认已在 [Issues](https://github.com/layui/layui/issues?q=is%3Aissue ) 中搜索过类似的问题，但没有找到相关解决方案。
- [X] 我已仔细阅读： 🍀[ Layui Issue 贡献指南](https://github.com/layui/layui/blob/main/CONTRIBUTING.md)

### 议题类型

功能请求

### 使用版本

v2.9.14

### 问题描述

如图，图1正常打开后，图2是我按了F12后，我想按了F12后，怎么自动调整高度呢？。。。
图3和图4是别人cms系统中的，我应该怎么写？
![x1](https://github.com/user-attachments/assets/0d5770f1-7802-4f8c-985e-1985fff60197)
![x2](https://github.com/user-attachments/assets/9d4cd713-0c60-4c49-b56c-573df53dd6eb)
![x3](https://github.com/user-attachments/assets/a1e1b4b9-8cd7-4b83-9f84-92337a67ee96)
![x4](https://github.com/user-attachments/assets/159c9e88-8c6c-4ab0-8271-b51a6a082e18)

### 业务代码

```auto
/**
 * 显示一个层
 * @param {string|null} title 层的标题,如果为空则不显示标题
 * @param {string} url 层要显示的内容的URL
 * @param {number|string|null} w 层的宽度,如果不设置,则默认为窗口宽度减去50px
 * @param {number|null} type 层的类型,决定层的显示样式,0为默认,1为从右滑入,其他值为默认
 */
function layerDrawer(title, url, w) {
    if (title == null || title == '') {
        title = false;
    }
    if (url == null || url == '') {
        layer.msg('url地址不能为空');
        return false;
    }
    if (w == null || w == '' || w === undefined) {
        var window_width = $(window).width();
        if (window_width < 700) {
            w = window_width;
        } else if (window_width >= 700 && window_width < 1200) {
            w = 700;
        } else {
            w = window_width * 0.55;
        }
    }
    width = w;
    height = $(window).height();
    // 生成层的唯一标识符
    var layer_id = 'global_layer_drawer';
    // 使用layer框架打开一个层
    layer.open({
        id: layer_id,
        type: 2,
        area: [width + 'px', height + 'px'],
        offset: 'r',
        // 从右往左
        anim: 'slideLeft',
        scrollbar: false,
        maxmin: false,
        shade: 0.4,
        resize: false,
        title: title,
        content: url
    });
}
```

### 截图补充

@Sight-wcg 晨星 大佬，出来帮忙看下。。。

### 浏览器

最新edge

### 友好承诺

- [X] 我承诺将本着相互尊重、理解和友善的态度进行交流，共同维护 Layui 良好的社区氛围。

**Comment by Sight-wcg** [comment]
```
area: [width + 'px', 100%],
```

我猜和高度无关，有两种可能
1. 截图中按钮在 iframe 内部，用 position: fixed 相对于视口定位到底部
2. 截图中按钮是通过修改 layer footer 实现的，类似下面的方法，仅供参考
```
    /** 绑定表单弹窗, iframe 要额外处理 */
    var modelForm = function (layero, btnFilter, formFilter) {
        var btnGroup = layero.find('.layui-layer-btn');
        var btnOk = btnGroup.find('.layui-layer-btn0');
        var btnReset = btnGroup.find('.layui-layer-btn1');

        layero.addClass('layui-form').attr('lay-filter', formFilter);
        //btnOk.attr('lay-submit', '').attr('lay-filter', btnFilter).prepend('<i class="layui-icon layui-icon-xxx"><i>');
        btnOk.prepend('<i class="layui-icon layui-icon-xxx"><i>');
        btnOk.on('click', function(){
          layui.form.submit(formFilter);
        })
          
        btnReset.prepend('<i class="layui-icon layui-icon-xxx"><i>');
        btnReset.on('click', function(){
          layui.form.render(null, formFilter)
        });
    };

**Reply by hulang** [reply]
使用1，可以了。`area: [width + 'px', 100%]`，我在想，`layer`是不是应该增加一个回调函数，当宽/高度 发生变化时，用户可以自由处理。。

**Reply by Sight-wcg** [reply]
> 使用1，可以了。`area: [width + 'px', 100%]`，我在想，`layer`是不是应该增加一个回调函数，当宽/高度 发生变化时，用户可以自由处理。。

最早设计是支持 IE6 的，所以「自由处理」相当繁琐。当前版本添加该回调意义不大，除非很熟悉 layer 源码，否则没什么用。

**Comment by LSL1618** [comment]
可以绑定`window.onresize`去处理页面宽高变化事件，然后借助`layer.style(index, css)`方法重置弹框宽高。

**Reply by hulang** [reply]
> 可以绑定`window.onresize`去处理页面宽高变化事件，然后借助`layer.style(index, css)`方法重置弹框宽高。

感谢回复，我直接使用：`area: [width + 'px', '100%']` 挺符合我所有项目场景，如果使用：`window.onresize`，就如我上面回复的，`layer`需要更新一个`onresize`回调函数，这样更加好的使用：`layer.style(index, css)`方法。

**Reply by LSL1618** [reply]
`area: [width + 'px', '100%']`只适合在页面宽高变化之后打开弹框有效，如果先打开弹框，再改变页面宽高，那就失效了。

**Reply by hulang** [reply]
> `area: [width + 'px', '100%']`只适合在页面宽高变化之后打开弹框有效，如果先打开弹框，再改变页面宽高，那就失效了。

我项目中，全部都的点击某个按钮进行弹出的，没有直接弹出的。。。

我还没有用过：`window.onresize`配合`layer.style(index, css)`，是不是当页面宽度和高度发生变化时，就去使用：`layer.style(index, css)`进行调整？问题是`index`索引，应该怎么获取呢？有的页面有不下5个弹出的，我上面有我弹出的函数，我应该怎么改进使用：`window.onresize`?

**Reply by LSL1618** [reply]
每次弹框open之后都会返回一个唯一的弹框索引，你肯定没仔细看过文档，这个你要用全局变量存起来，这样才能在layer.style()方法中引用。一般来讲一次只会操作一个弹框，很少会同时操作多个弹框的情况，这样就很好处理了。`window.onresize`事件绑定一次即可。
```
// 仅供参考
let dialogIndex = layer.open();
window.onresize = function () {
layer.style(dialogIndex, {
width: myWidth + 'px',
height: myHeight + 'px',
})
}
```

**Reply by hulang** [reply]
> 每次弹框open之后都会返回一个唯一的弹框索引，你肯定没仔细看过文档，这个你要用全局变量存起来，这样才能在layer.style()方法中引用。一般来讲一次只会操作一个弹框，很少会同时操作多个弹框的情况，这样就很好处理了。`window.onresize`事件绑定一次即可。
> 
> ```
> // 仅供参考
> let dialogIndex = layer.open();
> window.onresize = function () {
> layer.style(dialogIndex, {
> width: myWidth + 'px',
> height: myHeight + 'px',
> })
> }
> ```

谢谢。。明白了，我都是使用的我上面写好的函数，函数返回index，我有好几个页面中点击弹出的有8-10个，只需要定义一个`dialogIndex`变量，每次弹出，就得去8-10个取`dialogIndex`变量，到`window.onresize`判断`dialogIndex`是否`undefined`再进行调整。
